### PR TITLE
ORLP-195.Fix rating(back-end).

### PR DESCRIPTION
**Pull request by vldmr1703**
I added foreign keys for "deckId" and "courseId" in tables: "DeckRating" and "CourseRating.
Also you must rebuild tables.

### CODE REVIEW

File: src/main/java/com/softserve/academy/spaced/repetition/domain/CourseRating.java [hunk 1]
```diff
@@ -15,8 +15,9 @@
     @Column(name = "account_email", nullable = false)
     private String accountEmail;
 
-    @Column(name = "course_id", nullable = false)
-    private long courseId;
+    @OneToOne(cascade = CascadeType.ALL)
+    @JoinColumn(name = "course_id")
```
Comment (antonbakalets) [inline src/main/java/com/softserve/academy/spaced/repetition/domain/CourseRating.java]: I suppose one course can be rated by many users. one-to-many, isn't it?

Comment (vldmr1703) [reply src/main/java/com/softserve/academy/spaced/repetition/domain/CourseRating.java]: But with OneToOne all is working.
Logically it must work with ManyToOne, but I don't understand how it works.

Comment (vldmr1703) [reply src/main/java/com/softserve/academy/spaced/repetition/domain/CourseRating.java]: Maybe it is like refence key but is interpreted by idea as ManyToOne?

Comment (antonbakalets) [reply src/main/java/com/softserve/academy/spaced/repetition/domain/CourseRating.java]: I think I should work with @ ManyToOne(fetch = FetchType.LAZY) and possibly @JsonIgnore. And no mapping from Course to CourseRating.

Comment (vldmr1703) [reply src/main/java/com/softserve/academy/spaced/repetition/domain/CourseRating.java]: Ok, I will try it.

File: src/main/java/com/softserve/academy/spaced/repetition/DTO/impl/CardRatingPublicDTO.java [hunk 2]
```diff
@@ -15,7 +15,7 @@ public String getAccountEmail() {
     }
 
     public long getCardId() {
-        return getEntity().getCardId();
+        return getEntity().getCard().getId();
```
Comment (b0hdan) [inline src/main/java/com/softserve/academy/spaced/repetition/DTO/impl/CardRatingPublicDTO.java]: Why was this change required?

File: src/main/java/com/softserve/academy/spaced/repetition/domain/Deck.java [hunk 3]
```diff
@@ -31,12 +31,21 @@
     @JoinColumn(name = "user_id")
     private User deckOwner;
 
-    @ManyToMany(cascade = CascadeType.ALL)
+    @OneToMany(cascade = CascadeType.ALL)
     @JoinTable(name = "deck_cards", joinColumns = {
```
Comment (b0hdan) [inline src/main/java/com/softserve/academy/spaced/repetition/domain/Deck.java]: I don't think that the @JoinTable annotation is needed when an @OneToMany relationtship is used.

File: src/main/java/com/softserve/academy/spaced/repetition/repository/CardRatingRepository.java [hunk 4]
```diff
@@ -7,8 +7,8 @@
 
 public interface CardRatingRepository extends JpaRepository<CardRating, Long> {
 
-    @Query("select avg(c.rating) from CardRating c where c.cardId=:cardId group by c.cardId")
-    Double findRatingByCardId(@Param("cardId") long cardId);
+    @Query("select avg(c.rating) from CardRating c where c.card.id=:cardId group by c.card.id")
+    Double findRatingByCard_Id(@Param("cardId") long cardId);
```
Comment (b0hdan) [inline src/main/java/com/softserve/academy/spaced/repetition/repository/CardRatingRepository.java]: Why is this underline needed here?

File: src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java [hunk 5]
```diff
@@ -76,18 +76,18 @@ public void updateDeck(Deck updatedDeck, Long deckId, Long categoryId) {
...
+//        for(Course course: courses){
+//            decks=course.getDecks();
+//            for(Deck deck : decks){
+//                if (deck.getId().equals(deckId)) {
+//                    decks.remove(deck);
+//                    break;
+//                }
+//            }
+//            courseRepository.save(course);
+//        }
```
Comment (b0hdan) [inline src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java]: I'm sure, we don't need all of these comments here.

File: src/main/java/com/softserve/academy/spaced/repetition/DTO/impl/CardRatingPublicDTO.java (hunk 2 above)
Comment (vldmr1703) [reply src/main/java/com/softserve/academy/spaced/repetition/DTO/impl/CardRatingPublicDTO.java]: Because I added refence key to table CardRating and to get CardId we need call getCard() method.

File: src/main/java/com/softserve/academy/spaced/repetition/domain/Deck.java (hunk 3 above)
Comment (vldmr1703) [reply src/main/java/com/softserve/academy/spaced/repetition/domain/Deck.java]: You propose to save data about cards of some deck in table of decks?

File: src/main/java/com/softserve/academy/spaced/repetition/repository/CardRatingRepository.java (hunk 4 above)
Comment (vldmr1703) [reply src/main/java/com/softserve/academy/spaced/repetition/repository/CardRatingRepository.java]: Because here we have refence key.

File: src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java (hunk 5 above)
Comment (vldmr1703) [reply src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java]: Yesterday you said it is wrong to delete card this way.

File: src/main/java/com/softserve/academy/spaced/repetition/domain/Deck.java (hunk 3 above)
Comment (b0hdan) [reply src/main/java/com/softserve/academy/spaced/repetition/domain/Deck.java]: No, of course. no. But you don't need an additional table like "deck_cards" when using OneToMany relationship. In this case you have to use a foreign key in the table named Card.

File: src/main/java/com/softserve/academy/spaced/repetition/repository/CardRatingRepository.java (hunk 4 above)
Comment (b0hdan) [reply src/main/java/com/softserve/academy/spaced/repetition/repository/CardRatingRepository.java]: What does this underline mean? It's just a symbol, can it change something?

File: src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java (hunk 5 above)
Comment (b0hdan) [reply src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java]: Well, then it's better to either remove these comments or make them work correctly or leave them for now. But it's my opinion, I'm not sure. Answer to the previous comments better, please.

File: src/main/java/com/softserve/academy/spaced/repetition/repository/CardRatingRepository.java (hunk 4 above)
Comment (vldmr1703) [reply src/main/java/com/softserve/academy/spaced/repetition/repository/CardRatingRepository.java]: When we have foreign key, according to the arrangements we should name method so.
It is just convention.

File: src/main/java/com/softserve/academy/spaced/repetition/domain/Deck.java (hunk 3 above)
Comment (vldmr1703) [reply src/main/java/com/softserve/academy/spaced/repetition/domain/Deck.java]: I'll think about it and try to make changes.
